### PR TITLE
toolchain: Add macro for keeping an object in the binary

### LINKED ISSUE #34413: Improve __used attribute to actually keep requested function/variable
**Is your feature request related to a problem? Please describe.**
There is `__used` specified in `include/toolchain/gcc.h` which is setting attribute `used`. The idea is to keep marked function/variable in the binary even when it is not used. However, linker still removes it (thanks to garbage collection?). 

**Describe the solution you'd like**
Similarly as `__used`, `__func_used` could be added which adds code to dedicated section which is kept. Something like:
```
#define __func_used __attribute__((used)) __attribute__((section(".used_code")))
```

**Additional context**
So far for rpc `--no-gc-sections` was used to keep rpc function but that acts globally and generates issues as zephyr has this approach where api is available but code is not compiled in and relies on `IS_ENABLED`. See #26848.

### PR DESCRIPTION
**Pull request by nordic-krch**
Add KEEP_OBJECT macro which can be used to prevent linker garbage
collector from removing unused object (function or variable).

Example use case is to use it for functions which are part of a test
application which has functions which are indirectly called by the
host test tools. Since they are not called explicitely by the
application so without marking them with KEEP_OBJECT they would be
removed by the linker garbage collector.

Fixes #34413.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

**Comment by nordic-krch** [comment]
It was @tejlmand concern. Maybe he can elaborate:

https://github.com/zephyrproject-rtos/zephyr/issues/34413#issuecomment-823834066

> not convinced that the right thing to do here is to actually place all of those functions in same output section.
Of course it has the benefit to those users who don't care in which output section the code is placed.

I can image that there might be some esthetical objections or simple confusion. E.g. when inspecting memory map you cannot find your function under file section but somewhere else.

**Comment by andyross** [comment]
> I'm still a bit confused why directly placing the functions in the section instead of function pointers is a bad idea.

For any single class of functions for a single problem, sure.  But if you want a compile-time equivalent of a linker KEEP() statement, which is sort of what this API is, it's inevitably going to be used for something that wants some alternative special treatment in the linker and can't just be placed in its own section.

But definitely let's find a way to get those pointers out of the generated image.

File: include/zephyr/toolchain/common.h
Comment (andyross) [inline include/zephyr/toolchain/common.h:284]: More nitpickery: maybe "LINKER_KEEP()" or something would be better, to clarify that the macro does the same thing that the KEEP() expression does in a linker script?  The word "object" is ambiguous basically everywhere in software engineering.  Using "symbol" would be better if you want a noun IMHO, as that's the name of the the abstraction being managed.

Also, you can relax the documentation a little, the argument can be any pointer expression that points at compile time into linker-visible memory. `KEEP_OBJECT(global_object.field[7])` would work just fine to save the global_object.  (This is silly as written, but it's easy to imagine the argument itself being a macro that expands to some selected subobject or whatever).

Comment (nordic-krch) [reply include/zephyr/toolchain/common.h]: Renamed to `LINKER_KEEP` and use `symbol` term instead of object.

**Review by dcpleung — DISMISSED** [review]

**Review by stephanosio — DISMISSED** [review]

File: include/zephyr/linker/common-rom/common-rom-misc.ld
Comment (tejlmand) [reply include/zephyr/linker/common-rom/common-rom-misc.ld:22]: As second comment to this section, is there any reason why this is placed between:
```
	ITERABLE_SECTION_ROM(shell, 4)
```
and
```
SECTION_DATA_PROLOGUE(shell_root_cmds_sections,,)
```
breaking up the continues location of `shell` and `shell_root_cmds` sections ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
Generally looks good.
Couple of small comments and a single request.

Please ensure to add identical functionality here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/linker_script/common/common-rom.cmake

File: include/zephyr/linker/common-rom/common-rom-misc.ld
Comment (nordic-krch) [reply include/zephyr/linker/common-rom/common-rom-misc.ld]: thanks for spotting that. Fixed.

**Review by stephanosio — APPROVED** [review]

File: include/zephyr/linker/common-rom/common-rom-misc.ld
Comment (andyross) [inline include/zephyr/linker/common-rom/common-rom-misc.ld:25]: This is still in its own section internal to the image memory, so it's occupying flash or sram at runtime.   It shouldn't be hard to put this e.g. at the end of the link and pull it out with objcopy, etc...

Not worth -1, but I personally find the waste irksome.

Comment (carlescufi) [reply include/zephyr/linker/common-rom/common-rom-misc.ld:25]: @tejlmand ^^

Comment (nordic-krch) [reply include/zephyr/linker/common-rom/common-rom-misc.ld:25]: It can always be improved and won't change macro behavior. Current use case is for test applications so it does not matter that much.

Comment (tejlmand) [reply include/zephyr/linker/common-rom/common-rom-misc.ld:25]: I think @andyross has a very valid point.

I won't block this PR for it, but i'm concerned that if such improvement is left for the future it will simply be forgotten or down prioritized, so if you have the time now @nordic-krch then I believe a good time to do it would be now 😄

**Comment by tejlmand** [comment]
> > not convinced that the right thing to do here is to actually place all of those functions in same output section.
> > Of course it has the benefit to those users who don't care in which output section the code is placed.
> 
> I can image that there might be some esthetical objections or simple confusion. E.g. when inspecting memory map you cannot find your function under file section but somewhere else.
> 
> Imo, if there are any concerns putting a pointer is definitely harmless, especially that maybe it could be put in some `devnull` section that is stripped from binary (as @andyross suggested).

I believe this is a nice alternative, cause it keeps the original function (symbol) at it's expected location and just keeps a pointer to it.
Developers can then do this directly in code, not having to go through adding linker flags in CMake like `"-Wl,--undefined=<symbol>"`. A solution that itself is very ld-centric and thus not very portable.

@nordic-krch a 👍 for current proposal, and an extra 👍 if you manage to place it in a section that can be dropped later, like we do for the IDT_LIST:
https://github.com/zephyrproject-rtos/zephyr/blob/e0ce8436a584d281646da78eb7a12dffa6b44092/include/zephyr/linker/intlist.ld#L22-L31

**Review by tejlmand — APPROVED** [review]
lgtm.

But please let @nordic-krch  consider those comments before final merge:
https://github.com/zephyrproject-rtos/zephyr/pull/46841#issuecomment-1172013843
https://github.com/zephyrproject-rtos/zephyr/pull/46841#discussion_r911656349

**Pull request merged**
